src/config.h

```diff
@@ -319,3 +319,18 @@ void setcpuaffinity(const char *cpulist);
 #endif
 
 #endif
+
+/* Check if we can compile AVX2 code */
+#if defined (__x86_64__)
+    #if ((defined(__GNUC__) && __GNUC__ >= 7) || (defined(__clang__) && __clang_major__ >= 7))
+        #if defined(__has_attribute) && __has_attribute(target)
+            #define HAVE_AVX2
+        #endif
+    #endif
+#endif
+
+#if defined (HAVE_AVX2)
+    #define ATTRIBUTE_TARGET_AVX2 __attribute__((target("avx2")))
+#else
+    #define ATTRIBUTE_TARGET_AVX2
+#endif
```

Comment:
maybe I mislead you, since all your code is under the `HAVE_AVX2` macro definition, right? and could you show me some information about when gcc and clang start to support avx2

```suggestion
/* Check if we can compile AVX2 code */
#if defined (__x86_64__) && ((defined(__GNUC__) && __GNUC__ >= 7) || (defined(__clang__) && __clang_major__ >= 7))
#if defined(__has_attribute) && __has_attribute(target)
#define HAVE_AVX2
#endif
#endif

#endif
```

Reply:
The exact lower bounds of compiler versions are `x86-64 gcc 4.9.0` and `x86-64 clang 3.8` (tested from https://gcc.godbolt.org/z/vPPGj8GKb).

The asm is not optimal from gcc 5.1 to gcc 10.5, fixed in gcc 11.1. But it's still much faster than scalar version.

I just picked loose conditions to make it compile. I can change to the tight bounds if you like.

Reply:
This PR adds an AVX2 fast path. So all addtional code is under the `HAVE_AVX2` macro definition.

Reply:
so the below code is not necessary, you can use __attribute__((target("avx2")) since it is under the HAVE_AVX2 macro definition, right?

```
 #if defined (HAVE_AVX2)
    #define ATTRIBUTE_TARGET_AVX2 __attribute__((target("avx2")))
#else
    #define ATTRIBUTE_TARGET_AVX2
#endif
```

Reply:
`ATTRIBUTE_TARGET_AVX2` is unnecessary for this PR. 
I think it may be used in the future like `ATTRIBUTE_TARGET_POPCNT`.

Reply:
could we remove blank space just as other most macro definitions
```
#ifdef HAVE_AVX2
#define ATTRIBUTE_TARGET_AVX2 __attribute__((target("avx2")))
#else
#define ATTRIBUTE_TARGET_AVX2
#endif
```

Reply:
> could we remove blank space just as other most macro definitions

changed in 7cd8532a8f6bce93eaf8f7d8244ec65ef653f7db